Full rewrite and TypeScript migration

fmalcher opened the pull request:
This PR aims to fully restructure the module and add a bunch of new features:

- migration to TypeScript (resolve #12)
- cleaner file and folder structure, inspired by other modules
- instance feedback (resolve #3)
- timed transitions (resolve #6)
- relative fader changes (resolve #14)
- media player modes (resolve #16)
- media player shuffle feedback (resolve #23)
- 2-track recording actions and feedback (resolve #15)
- mute groups (resolve #22)



## TODO

- [x] actions for fader levels
- [x] feedback for AUX mute
- [x] update script
- [x] track feedback subscriptions (resolve #18)
- [x] stereo link (blocked in connection library, to be resolved there first)

fmalcher commented:
https://github.com/bitfocus/companion-module-soundcraft-ui/issues/12

fmalcher commented:
@Julusian Ready for first review! 😎

fmalcher commented:
Without the intention to put you under pressure – when do you think you could review this one here? 🙂 @Julusian Thanks!

Julusian commented on src/state.ts:
```diff
@@ -0,0 +1,69 @@
+   * @param stream$ The observable stream of values for this feedback
+   */
+  connect(evt: CompanionFeedbackEvent, stream$: Observable<any>): void {
+    const unsubscribe$ = this.feedbackUnsubscribe$.pipe(filter(fid => fid === evt.id || !fid));
```
It looks like calling connect multiple times for the same observable and evt.type will cause checkFeedbacks() to be called multiple times for each change.
This wont be good for performance

fmalcher replied:
Good catch! I think, this will only happen if someone registers the exact same feedback source on multiple banks in parallel.

Do you have an idea how to solve this? Debouncing would delay the result …

fmalcher replied:
I could do a very short debounce, though, to catch event bursts, like 10 ms

Julusian replied:
Yes, it would only be if using the same feedback source, but I dont think that unlikely. For example, I have the same mute button with feedback on 3 pages of my streamdeck (x32 not this module), which mutes my mic (for videocalls).

I would suggest tracking the active subscriptions, similar to what I have done here: https://github.com/bitfocus/companion-module-behringer-x32/blob/master/src/state.ts#L31
You will need some 'stream' id, but once you have that it will be easy to subscribe once and reference count the subscription

fmalcher replied:
I will put this into an issue and plan a rework after the major release!

fmalcher replied:
https://github.com/bitfocus/companion-module-soundcraft-ui/issues/18

Julusian reviewed (approved):
A couple of small tweaks, but generally looks good :)

fmalcher commented:
- added all features I wanted to have in this rework
- tidy up
- added migrations cript (and tested it 😉 )

From my perspective, this PR is ready to merge!
@Julusian @krocheck

fmalcher commented:
### ⚠️ Please rebase+merge instead of merge to keep the history clean 😉

![Screen Shot 2020-11-29 at 22 53 52](https://user-images.githubusercontent.com/1683147/100554572-ca457e80-3295-11eb-8100-182eacf22048.png)

fmalcher commented:
## ⚠️ Please rebase+merge instead of merge to keep the history clean 😉

![Screen Shot 2020-11-29 at 22 53 52](https://user-images.githubusercontent.com/1683147/100554572-ca457e80-3295-11eb-8100-182eacf22048.png)

krocheck commented:
@Julusian are you still happy here? Seems to be ready to go.

Julusian commented:
@krocheck I am happy for it to be merged. I haven't done another review, but I don't think that is necessary
